torch/csrc/jit/codegen/cuda/type.cpp

```diff
@@ -71,6 +95,21 @@ DataType promote_type(const DataType& t1, const DataType& t2) {
       t1,
       " and ",
       t2);
+  // FIXME: type promotion is not as simple as (t1 < t2 ? t1 : t2)
+  // hint:
+  // half + bfloat = float
+  // double + complex float = complex double
```

Comment:
❤️ any help you want to give us here is greatly appreciated. Type promotion is getting better but there seems to always be some gaps. You could check our python tests to see what's supported in test_jit_cuda_fuser.py feel free to add disabled tests and file issues against us to, can have someone fix it if you can tell us what's broken ;-)

Reply:
@zasdfgbnm  This is not the only place we handle type promotion.  I believe the first case on the hint is handled here:
[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/type_promotion.cpp#L170-L173
](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/type_promotion.cpp#L170-L173
)

and the second case should work if we lift this check since we are trying to do exactly the same thing eager mode does
[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/type_promotion.cpp#L55-L57](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/type_promotion.cpp#L55-L57)

But still please feel free to file issues if any case breaks.  Would really appreciate that. 😀

Reply:
@shmsong Interesting that there are two implementations of the type promotion logic. Thanks for the info! I will do a more thorough reading of the code about type promotion, and file PRs when necessary.